In [107]:
import pandas as pd

df = pd.read_csv("sales_data.csv")
df.head()

order_id     customer_name   region           product   sales_rep  \
0  ELX-1000          TechMart    North   Fitbit Charge 6    Jane Doe   
1  ELX-1001          TechMart  Central        Power Bank  John Smith   
2  ELX-1002  Electronics Plus  Central        Power Bank    Jane Doe   
3  ELX-1003  Electronics Plus     West  Screen Protector  Bob Wilson   
4  ELX-1004        TechMart       East  Wireless Charger  Mike Davis   

   order_date  quantity  unit_price discount  shipping_cost    revenue  \
0  2023-04-25         9      887.72     0.17          22.84  $6,631.27   
1  2023-01-16         6      274.09     0.13           5.53    1430.75   
2  2023-08-18        38      344.66     0.22          20.18   10215.72   
3  2023-03-21        14     1149.30     0.08           6.85   14802.98   
4  2023-05-16         3      879.73     0.13          24.46     2296.1   

   total_cost priority      sales_rep_email  
0     6654.11   medium    janedoe@gmail.com  
1     1436.28     HIGH  JOHNSMITH@YAHOO.COM  
2    10235.90     High    janedoe@gmail.com  
3    14809.83   medium  bobwilson@gmail.com  
4     2320.56   Medium     mikedavis@avc.io

In [108]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1050 entries, 0 to 1049
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_id         1050 non-null   object 
 1   customer_name    1050 non-null   object 
 2   region           1050 non-null   object 
 3   product          1050 non-null   object 
 4   sales_rep        1050 non-null   object 
 5   order_date       1050 non-null   object 
 6   quantity         1050 non-null   int64  
 7   unit_price       1050 non-null   float64
 8   discount         1029 non-null   object 
 9   shipping_cost    1050 non-null   float64
 10  revenue          1016 non-null   object 
 11  total_cost       1050 non-null   float64
 12  priority         918 non-null    object 
 13  sales_rep_email  7 non-null      object 
dtypes: float64(3), int64(1), object(10)
memory usage: 115.0+ KB


Looking at the DataFrame info, several columns have incorrect data types that should be converted:
Columns that need conversion:

*order_date (currently object)* - Should be converted to datetime64 since this represents dates  
*discount (currently object)* - Should be converted to float64 since discounts are typically numeric values (percentages or amounts)  
*revenue (currently object)* - Should be converted to float64 since revenue is a monetary/numeric value.

*customer_name*

In [109]:
df.customer_name.unique()

array(['TechMart', 'Electronics Plus', '  TechMart  ', 'GadgetHub',
       'Digital World', 'Smart Store', '  Smart Store  ',
       '  Digital World  ', '  GadgetHub  ', '  Electronics Plus  ',
       '    Digital World    ', '    Smart Store    ',
       '      Electronics Plus      ', '  Smart Store   Inc.',
       '  TechMart   Inc.', 'GadgetHub Inc.', 'Electronics Plus Inc.',
       'TechMart Inc.', 'Digital World Inc.', 'Smart Store Inc.',
       '  Electronics Plus   Inc.', '  GadgetHub   Inc.'], dtype=object)

In [110]:
import re

df['customer_name'] = (
    df['customer_name']
    # Remove leading and trailing spaces
    .str.strip() 

    # Replace multiple spaces with a single space
    .apply(lambda x: re.sub(r'\s+', ' ', x))  

    # Remove 'Inc.', 'Inc', and variations with trailing spaces
    .str.replace(r'\s*Inc\.*$', '', regex=True)
    
    # Convert to title case for consistency 
    .str.title()  
)

In [111]:
df.customer_name.unique()

array(['Techmart', 'Electronics Plus', 'Gadgethub', 'Digital World',
       'Smart Store'], dtype=object)

*region*

In [112]:
df.region.unique()

array(['North', 'Central', 'West', 'East', 'South', 'west', 'east',
       'central', 'NORTH', 'WEST', 'EAST', 'CENTRAL', 'north', 'south',
       'SOUTH'], dtype=object)

In [113]:
df['region']=df.region.str.capitalize()

In [114]:
df.region.unique()

array(['North', 'Central', 'West', 'East', 'South'], dtype=object)

*product*

In [115]:
df['product'].value_counts().sort_index()

product
AIRPODS PRO              2
AirPods Pro             50
AirPods-Pro              5
Apple AirPods Pro        2
Apple MacBook Pro        1
Apple Watch Series 9    62
Apple iPhone 15          5
Bluetooth Speaker       61
Dell XPS 13             62
Fitbit Charge 6         62
Galaxy S24               2
Gaming Headset          54
IPHONE 15                1
IPHONE CASE              2
Laptop Stand            68
MACBOOK PRO              2
MacBook Pro             43
MacBook-Pro              1
Phone Case               1
Phone Holder            59
Portable Charger         2
Power Bank              47
PowerBank                2
SAMSUNG GALAXY S24       3
SCREEN PROTECTOR         1
Samsung  Galaxy  S24     3
Samsung Galaxy S24      54
Screen Protector        46
Screen-Protector         3
Sony WH-1000XM5         55
Type-C Cable             3
USB C Cable              6
USB-C Cable             42
USBC Cable               1
WIRELESS CHARGER         3
Wireless Charger        78
Wireless-Charger    

*order_date*

In [116]:
# Convert order_date to datetime
df['order_date'] = pd.to_datetime(df['order_date'])

*discount*

In [117]:
df.discount

0       0.17
1       0.13
2       0.22
3       0.08
4       0.13
        ... 
1045    0.03
1046    0.11
1047    0.08
1048    0.03
1049    0.25
Name: discount, Length: 1050, dtype: object

The discount column contains a mix of numeric values like 0.12 and 0.1, along with some nulls and "-" as placeholders. Since it's currently stored as an object type (string), which is not appropriate for numeric operations, we'll convert it to a proper numeric type (float).

In [118]:
# Convert discount to float
df['discount'] = pd.to_numeric(df['discount'], errors='coerce')

*revenue*

In [119]:
df.revenue

0       $6,631.27
1         1430.75
2        10215.72
3        14802.98
4          2296.1
          ...    
1045     10922.39
1046     31138.37
1047      1688.18
1048      4458.65
1049      7020.29
Name: revenue, Length: 1050, dtype: object

The revenue column is currently of type object and contains values like "$6,631.27", which can't be directly converted to numeric due to the dollar sign and comma.

So before converting it to a numeric type, we’ll first remove the leading dollar sign  and comma.

In [120]:
# Remove the leading dollar sign from the 'revenue' values
df['revenue'] = df['revenue'].str.lstrip('$')

# Remove commas from the 'revenue' values to prepare for numeric conversion
df['revenue'] = df['revenue'].str.replace(',', '', regex=False)

In [121]:
# Convert to numeric type
df['revenue'] = pd.to_numeric(df['revenue'], errors='coerce')

*priority*

In [122]:
df['priority'].unique()

array(['medium', 'HIGH', 'High', 'Medium', nan, 'Low', 'high', 'low'],
      dtype=object)

In the priority column, we have values with inconsistent capitalization.
To make the values consistent, we'll standardize the text format, so that all values follow the same casing (e.g., 'High', 'Medium', 'Low').

In [123]:
df.priority=df.priority.str.capitalize()

Furthermore, we can convert the priority column into a pandas Categorical data type with a defined order. This helps pandas understand the logical sequence of priorities (i.e., 'Low' < 'Medium' < 'High') for sorting and comparisons.

In [124]:
priority_order = ['Low', 'Medium', 'High']

df['priority'] = pd.Categorical(
    df['priority'],
    categories=priority_order,
    ordered=True
)

In [125]:
df.priority.sort_values()

1049    Low
269     Low
706     Low
266     Low
265     Low
       ... 
1017    NaN
1018    NaN
1025    NaN
1032    NaN
1038    NaN
Name: priority, Length: 1050, dtype: category
Categories (3, object): ['Low' < 'Medium' < 'High']

*sales_rep_email*

In [126]:
# Convert all email addresses in 'sales_rep_email' column to lowercase
df['sales_rep_email'] = df.sales_rep_email.str.lower()

In [127]:
# Returns rows where sales_rep_email is NULL
df[df['sales_rep_email'].isnull()]

order_id     customer_name   region               product   sales_rep  \
6     ELX-1006          Techmart    North  Apple Watch Series 9  Bob Wilson   
7     ELX-1007     Digital World    South      Wireless Charger  Bob Wilson   
8     ELX-1008       Smart Store    South          Phone Holder    Jane Doe   
9     ELX-1009       Smart Store    South      Screen Protector  John Smith   
11    ELX-1011       Smart Store  Central       Fitbit Charge 6  Mike Davis   
...        ...               ...      ...                   ...         ...   
1045  ELX-1764     Digital World     West          Laptop Stand  Mike Davis   
1046  ELX-1511     Digital World     West        Gaming Headset  Mike Davis   
1047  ELX-1612     Digital World    South      screen protector  Bob Wilson   
1048  ELX-1542  Electronics Plus    South       Sony WH-1000XM5    Jane Doe   
1049  ELX-1806  Electronics Plus    North      Wireless Charger    Jane Doe   

     order_date  quantity  unit_price  discount  shipping_cost   revenue  \
6    2023-02-10        15     1041.78      0.10          14.07  14064.03   
7    2023-04-18        43      331.37      0.23          17.96  10971.66   
8    2023-03-25        30      464.65      0.25          17.80  10454.62   
9    2023-04-28         3      968.98       NaN           6.32   2616.25   
11   2023-08-08        38      488.29      0.05          24.95  17627.27   
...         ...       ...         ...       ...            ...       ...   
1045 2023-08-16        30      375.34      0.03          20.79  10922.39   
1046 2023-01-04        33     1060.21      0.11           9.60  31138.37   
1047 2023-11-01        42       43.69      0.08          24.75   1688.18   
1048 2023-11-13        23      199.85      0.03          23.40   4458.65   
1049 2023-06-23        13      720.03      0.25           7.46   7020.29   

      total_cost priority sales_rep_email  
6       14078.10      Low             NaN  
7       10989.62     High             NaN  
8       10472.42   Medium             NaN  
9        2622.57   Medium             NaN  
11      17652.22     High             NaN  
...          ...      ...             ...  
1045    10943.18      Low             NaN  
1046    31147.97     High             NaN  
1047     1712.93      Low             NaN  
1048     4482.05   Medium             NaN  
1049     7027.75      Low             NaN  

[1043 rows x 14 columns]

The sales_rep_email column contains some null values. To fill them based on matching sales_rep names, we first create a mapping and then converts the filtered DataFrame into a dictionary.

In [ ]:
email_map = df[df['sales_rep_email'].notnull()].drop_duplicates('sales_rep')[['sales_rep', 'sales_rep_email']]
email_dict = dict(zip(email_map['sales_rep'], email_map['sales_rep_email']))

In [131]:
email_map

sales_rep          sales_rep_email
0      Jane Doe        janedoe@gmail.com
1    John Smith      johnsmith@yahoo.com
3    Bob Wilson      bobwilson@gmail.com
4    Mike Davis         mikedavis@avc.io
10  Alice Brown  alice_browny@company.ai

In [132]:
email_dict

{'Jane Doe': 'janedoe@gmail.com',
 'John Smith': 'johnsmith@yahoo.com',
 'Bob Wilson': 'bobwilson@gmail.com',
 'Mike Davis': 'mikedavis@avc.io',
 'Alice Brown': 'alice_browny@company.ai'}

In [133]:
# Fill missing emails by mapping from sales_rep
df['sales_rep_email'] = df['sales_rep_email'].fillna(df['sales_rep'].map(email_dict))